### ISSUE TITLE
handling oauth redirect callbacks

### ISSUE THREAD

**Issue by mcku** [aws-amplify/amplify-flutter#2681]
### Description

Hi, I am using the dev preview package of amplify flutter cognito next-5. The redirects work fine on android, however, when configured for the web as an additional entry, let's say `https://www.myapp.com/` , also using auth.myapp.com as the cognito custom domain, the social oauth provider returns a callback with a `code` query parameter, however the parameter doesn't get used by the Amplify.Auth and hence authentication never completes (for the web case only). 

`Amplify.Auth.signInWithWebUi(provider:..)` initiates the login as a browser redirect (not a popup) and receives another redirect after login. But that last redirect I believe reloads the app. I think this is not supposed the way it should behave. 
With the code received, of course one can exchange the code for a token. But this is a labor intensive task. (https://aws.amazon.com/blogs/mobile/understanding-amazon-cognito-user-pool-oauth-2-0-grants/). 

Please advise,  when configuring for the flutter web, is the code-to-token flow necessary or is something apparently missing in my setup? 

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [ ] Authenticator
- [ ] DataStore
- [ ] Storage

### Steps to Reproduce

1. add auth redirect urls for the web 
  https://www.myapp.com/ both for sign in and sign out
2. use social provider login
3. login is not successful

this can be triggered by cognito hosted ui through aws console as well. 

### Platforms

- [ ] iOS
- [ ] Android
- [X] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.7.0

### Amplify Flutter Version

1.0.0-next.5

### Deployment Method

Amplify CLI

**Comment by dnys1** [comment]
Hi @mcku - sorry you are facing this issue. The Amplify Flutter Auth library is set up to fully support the OAuth 2 protocol on Web such that you do not need to handle any component yourself. However, there are some nuances when integrating with Flutter itself which could be causing issues.

The redirect to the social provider and back to the app is expected. We use the Authorization Code grant, which is the most secure option for frontend libraries.

Can you please share the following information?
1. The sign in URLs you've configured for your app
2. The URL where your app is hosted. Is it hosted at the root of your domain or at a subpath? e.g. `https://myapp.com/` vs. `https://myapp.com/app/`
3. How are you currently hosting your app?

Thank you

**Comment by mcku** [comment]
Hi @dnys1, 
Thanks for helping out! The answers you have asked for are:

1) The sign in URLs configured by the CLI
```myapp://,https://beta.myapp.com/,http://localhost:3000/```

2) The app is hosted at `https://beta.myapp.com/` at the root. 

3) The app is hosted by amplify hosting. There's a third party DNS involved, which maps `beta.myapp.com` CNAMEs to amplify hosting `<id>.cloudfront.net`. 
Also the amplify frontend branch is associated with the `beta.myapp.com` domain.

**Comment by dnys1** [comment]
Thanks for the information, @mcku. I will try to reproduce and see what the cause could be.

When your app starts after the redirect, do you see any errors in the console? Would you be able to turn on verbose logging and share the output here? To do so:

```dart
void main() {
  AWSLogger().logLevel = LogLevel.verbose;

  //... the rest the same
}

**Comment by mcku** [comment]
@dnys1 

I've enabled the verbose logs. As far as I could see, there is nothing different in the verbose logs. After redirection, the app reload itself. On the splash screen, I check for `Amplify.Auth.fetchSession()`, which yields not signed in, for this case. As I said above, the custom router catches the code and state params. But the reloaded app's session remains unauthenticated.

BTW the app uses Authenticator UI components and also has its own custom router, just for the purpose of handling query parameters. I've enabled auth persistence with indexedDb, in case it might recover from there, but did not help.

I am not sure if there are any assumptions/restrictions about the use  of a custom router. How is query parameter handling by flutter web authenticator component supposed to work?

**Comment by dnys1** [comment]
Understood, thanks @mcku. Can you try moving your Amplify configuration to your `main` method, like below, so that it runs before the router loads?

```dart
Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();

  // Same as before just moved to `main`
  await Amplify.addPlugins([...]);
  await Amplify.configure(amplifyconfig);

  runApp(...);
}
```

**Comment by dnys1** [comment]
Hi @mcku - I was able to reproduce this issue and believe I have a fix in #2651. I am working to get this merged either today or tomorrow and will do a release soon after. Thanks for your patience!

**Comment by mcku** [comment]
@dnys1 Thanks. I can confirm that the issue is now resolved.
